File: Java/MergeTwoBinaryTrees.java
/**
 * Created by Ethan on 2017/9/17.
 */
public class MergeTwoBinaryTrees {
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        if (t1 == null && t2 == null){
            return null;
        }
        int val = (t1 != null ? t1.val : 0) + (t2 != null ? t2.val : 0);
        TreeNode node = new TreeNode(val);
        node.left = mergeTrees(t1 != null ? t1.left : null, t2 != null ? t2.left : null);
        node.right = mergeTrees(t1 != null ? t1.right : null, t2 != null ? t2.right : null);

        return node;
    }
}

File: SwiftCode/MergeTwoBinaryTrees.swift
//
//  MergeTwoBinaryTrees.swift
//  SwiftCode
//
//  Created by Luminoid Li on 1/7/23.
//

import Foundation

//@main
private class Solution {
    // Time complexity: O(m + n)
    // Space complexity: O(m + n)
    func mergeTrees(_ root1: TreeNode?, _ root2: TreeNode?) -> TreeNode? {
        if root1 == nil && root2 == nil {
            return nil
        }
        let root = TreeNode()
        root.val = (root1?.val ?? 0) + (root2?.val ?? 0)
        root.left = mergeTrees(root1?.left, root2?.left)
        root.right = mergeTrees(root1?.right, root2?.right)
        return root
    }
    
    static func main() {
        let solution = Solution()
    }
}
